File: cuerda.py
import os
import math
import time

g = 9.81        # Gravedad (m/s^2)
dt = 0.01       # Intervalo de tiempo (s)
stiffness = 0.9 # Rigidez de la cuerda
numParticles = 10  # Número de partículas en la cuerda
segmentLength = 2.0 # Longitud entre partículas

class Particle:
    def __init__(self):
        self.x = 0
        self.y = 0
        self.vx = 0
        self.vy = 0
        self.fixed = False

def aplicarGravedad(p):
    if (p.fixed == False):
        p.vy += g * dt;
    
def actualizarParticula(p):
    if (p.fixed == False):
        p.x += p.vx * dt;
        p.y += p.vy * dt;
    
def aplicarRestriccion(p1, p2):
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    L = math.sqrt(dx * dx + dy * dy)
    diff = L - segmentLength
    a = (dx / L) * diff
    b = (dy / L) * diff

    if (p2.fixed == False):
        p2.x -= a
        p2.y -= b

    # dist = math.sqrt(dx * dx + dy * dy)
    # diff = (dist - segmentLength) / dist
    # if (p1.fixed == False):
        # p1.x += stiffness * diff * dx
        # p1.y += stiffness * diff * dy
    # if (p2.fixed == False):
        # p2.x -= stiffness * diff * dx
        # p2.y -= stiffness * diff * dy

def dibujarCuerda(particles):
    os.system("cls")
    for j in range(20):
        for i in range(40):
            particleFound = False
            for p in particles:
                if (int(p.x) + 10 == i and int(p.y) == j):
                    print("O", end="")
                    particleFound = True
                    break
            
            if (particleFound == False):
                print(".", end="")
        
        print()
    

particles = []

# Inicializamos la cuerda en posición horizontal
for  i in range(numParticles):
    part = Particle()
    part.x = i * segmentLength
    part.y = 0
    part.fixed = i == 0
    particles.append(part)
    

while (True):
    # Aplicamos la gravedad y actualizamos las partículas
    for p in particles:
        aplicarGravedad(p);
        actualizarParticula(p);

    # Aplicamos restricciones de distancia entre partículas adyacentes
    for i in range(numParticles - 1):
        aplicarRestriccion(particles[i], particles[i + 1]);

    # Dibujamos la cuerda
    dibujarCuerda(particles);

    # Pausa
    time.sleep(0.001)

File: cuerda.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <chrono>
#include <thread>

const double g = 9.81;        // Gravedad (m/s^2)
const double dt = 0.01;       // Intervalo de tiempo (s)
const double stiffness = 0.1; // Rigidez de la cuerda
const int numParticles = 10;  // Número de partículas en la cuerda
const double segmentLength = 1.0; // Longitud entre partículas

struct Particle {
    double x, y;    // Posición de la partícula
    double vx, vy;  // Velocidad de la partícula
    bool fixed;     // Si la partícula está fija (en un punto de anclaje)

    Particle(double posX, double posY, bool isFixed = false)
        : x(posX), y(posY), vx(0), vy(0), fixed(isFixed) {}
};

void aplicarGravedad(Particle &p) {
    if (!p.fixed) {
        p.vy += g * dt;
    }
}

void actualizarParticula(Particle &p) {
    if (!p.fixed) {
        p.x += p.vx * dt;
        p.y += p.vy * dt;
    }
}

void aplicarRestriccion(Particle &p1, Particle &p2) {
    double dx = p2.x - p1.x;
    double dy = p2.y - p1.y;
    double dist = std::sqrt(dx * dx + dy * dy);
    double diff = (dist - segmentLength) / dist;

    if (!p1.fixed) {
        p1.x += stiffness * diff * dx;
        p1.y += stiffness * diff * dy;
    }
    if (!p2.fixed) {
        p2.x -= stiffness * diff * dx;
        p2.y -= stiffness * diff * dy;
    }
}

void dibujarCuerda(const std::vector<Particle> &particles) {
    system("clear"); // Limpia la pantalla (en Windows usa system("cls"))
    for (int j = 0; j < 20; j++) {
        for (int i = 0; i < 40; i++) {
            bool particleFound = false;
            for (const auto &p : particles) {
                if (static_cast<int>(p.x) == i && static_cast<int>(p.y) == j) {
                    std::cout << "O";
                    particleFound = true;
                    break;
                }
            }
            if (!particleFound) std::cout << ".";
        }
        std::cout << std::endl;
    }
}

int main() {
    std::vector<Particle> particles;

    // Inicializamos la cuerda en posición horizontal
    for (int i = 0; i < numParticles; i++) {
        particles.emplace_back(i * segmentLength, 0, i == 0); // La primera partícula está fija
    }

    while (true) {
        // Aplicamos la gravedad y actualizamos las partículas
        for (auto &p : particles) {
            aplicarGravedad(p);
            actualizarParticula(p);
        }

        // Aplicamos restricciones de distancia entre partículas adyacentes
        for (size_t i = 0; i < particles.size() - 1; i++) {
            aplicarRestriccion(particles[i], particles[i + 1]);
        }

        // Dibujamos la cuerda
        dibujarCuerda(particles);

        std::this_thread::sleep_for(std::chrono::milliseconds(50)); // Pausa
    }

    return 0;
}
